### ISSUE TITLE
Add type hints

### ISSUE THREAD

**Issue by leakec** [leakec/tfc#9]
## Feature
I've been lagging on type hints since they are still changing in different Python versions. However, recently JAX has dropped support for numpy 1.19, and the `numpy.typing.NDArray` seems to play well with JAX arrays. Opening this to start looking into type hints.

**Comment by leakec** [comment]
Keeping this list for types:
* Use `numpy.typing.NDArray` for numpy arrays. This works with JAX and numpy.
* Use `Union[int, float, complex]` or `Union[int, float]` for generic number types. The `numbers.Number` works fine with `mypy`, but linting can sometimes throw errors. If the `Union` versions are used, everything works well with `mypy` and with linting.

**Comment by leakec** [comment]
Similar to #8, I'm pretty busy at work at the moment, so this issue will be tackled slowly.

**Comment by leakec** [comment]
Created `type_hints` branch for this issue.

**Comment by leakec** [comment]
Files to be type hinted:
* [ ] mtfc.py
* [ ] utfc.py
* [ ] utils/BF/BF_Py.py
* [ ] utils/PlotlyMakePlot.py
* [ ] utils/TFCUtils.py
* [ ] utils/MayaviMakePlot.py
* [ ] utils/Html.py
* [ ] utils/MakePlot.py
* [ ] utils/Latex.py

**Comment by leakec** [comment]
I'm going to skip types for some of the internal functions of `_SetupJax`. The types here are strange since they are used for JAX primitives, and no end-user will ever see these types since they never interact with these functions directly. Therefore, I see no harm in skipping them.

**Comment by leakec** [comment]
As of https://github.com/leakec/tfc/commit/f2167d4bbc5611db1462befbc087f7a2a46129d2 type hints have been added everywhere I believe they may be needed. I'm going to wait to close this issue until after I do some testing.

I'd like to open quite a few of the examples with an LSP and make tweaks based on the errors it thinks it spots.

**Comment by leakec** [comment]
Okay, I've mypy-ed around a bit and am satisfied. Pyright is still giving some warnings/errors at certain points, but these mostly seem to be the way it is dealing with the union of types that the jax arrays output. I'm not concerned with these, and am okay to merge at this point. This may require some tweaking in the future, but is good enough to close the issue.

**Comment by leakec** [comment]
Closing; https://github.com/leakec/tfc/pull/12 completes this.

**Comment by leakec** [comment]
Info in an email for version 1.0:

*TL;DR*
* Recently added support for generic field types and LSP support. These are the last major features I had on my to-add list. This means we can now do things like solve differential equations on complex domains, and you should get a lot more assistance when using the TFC package in your favorite text editor (provided it has LSP support, which it most likely does). 
* I added a Git runner and updated all of the examples so they work with the newest versions of TFC/JAX (except the huge deep-learning ones, since they take up too much memory/time for the free Git servers, but this is only approximately 3 tests).
* The TFC package is now going into "maintenance" mode. This means I'll keep up with parent package changes, and will add support for new OSs / Python versions as they do. I'm also happy to tackle feature requests that others may have, and of course, I'm always around to fix anything that breaks or any issues you may find.

*The long version:*
Updates earlier this year to add PEJIT (partial-evaluation just-in-time compilation) allow us to evaluate and cache basis functions at compile time rather than evaluating them in the optimization function. This opens up tons of possibilities since we don't need XLA transforms for the basis functions. Examples include (and I have tested):
Running problems on the GPU rather than the CPU.
Creating a Python-based backend so we can support more number types than double-precision floats.
Running things faster, since we don't need to compute the basis functions at run-time. In practice, this increase is negligible, since the basis functions were never the bottleneck.
This opens up other examples that I haven't tested, such as running things on a TPU. However, the main, recent change I pursued was a Python-based basis function backend to support different numeric types. This addition means we can now do things like solve differential equations on complex domains (and there is an example for that included in the documentation).

I've also added typing support to the package. This means that the interactions with LSP servers have improved noticeably, and you get a lot more useful feedback from the LSP if you're using one. This isn't perfect yet, but most of the imperfections are upstream (in JAX or NumPy), and as they make future improvements to their typing systems, I can make adjustments to make this even better.

I've added some tutorials for these new features as well. I've also added tutorials/lessons on PEJIT and some of the other advanced tools that the TFC package offers.

Those (Python-based basis functions and LSP) were the last two features I had on my to-add list, so I felt justified to move it things out of beta and into a full release. Before I did, I cleaned up all (everything except for the super time-consuming deep-learning stuff) the examples, and ensured they all worked. I also created a Git runner to run all of these examples, so it is easy to ensure things are working in the future.  

I don't have any large future plans for the package at this point. I'm planning to maintain support as new JAX versions come up, and add binary wheel releases for any new Python versions and OSs that they decide to support. The only thing I really see on the horizon is a potential Julia port whenever the [diffractor package](https://github.com/JuliaDiff/Diffractor.jl) is finished. This would offer more automatic differentiation support than anything I know of in Python, and Julia's JIT could reach into the source code I write, so there may be some performance gains (relative to Python) as well. Until then, I have no major updates planned. Of course, I'll still be around to fix anything or help with any issues that come up.

If you'd like more details on any of this feel free to check out the documentation (and if you want a lot of detail check out the relevant issues).
